packages/state-transition/src/util/loadState.ts

```diff
@@ -0,0 +1,251 @@
+import {CompositeTypeAny, Type} from "@chainsafe/ssz";
+import {ssz} from "@lodestar/types";
+import {ForkSeq} from "@lodestar/params";
+import {ChainForkConfig} from "@lodestar/config";
+import {BeaconStateAllForks, BeaconStateAltair, BeaconStatePhase0} from "../types.js";
+import {VALIDATOR_BYTES_SIZE, getForkFromStateBytes, getStateTypeFromBytes} from "./sszBytes.js";
+
+type BeaconStateType =
+  | typeof ssz.phase0.BeaconState
+  | typeof ssz.altair.BeaconState
+  | typeof ssz.bellatrix.BeaconState
+  | typeof ssz.capella.BeaconState
+  | typeof ssz.deneb.BeaconState;
+
+type MigrateStateOutput = {state: BeaconStateAllForks; modifiedValidators: number[]};
+
+/**
+ * Load state from bytes given a seed state so that we share the same base tree. This gives some benefits:
+ * - Have single base tree across the application
+ * - Faster to load state
+ * - Less memory usage
+ * - Ultilize the cached HashObjects in seed state due to a lot of validators are not changed, also the inactivity scores.
+ * @returns the new state and modified validators
+ */
+export function loadState(
+  config: ChainForkConfig,
+  seedState: BeaconStateAllForks,
+  stateBytes: Uint8Array
+): MigrateStateOutput {
```

Comment:
Consider moving the details of type manipulation to the package types/utils/container.ts

```ts
  const migratedState = deserializeContainerIgnoreFields(stateBytes, ["validators", "inactivityScores"]);

  // validators is rarely changed
  const validatorsBytes = sliceContainerField(stateType, "validators", stateBytes)
  const modifiedValidators = loadValidators(
    migratedState,
    seedState,
    validatorsBytes,
  );
```